fiatconnect-api.md

```diff
@@ -383,105 +406,245 @@ The FiatConnect API specifies *two* types of authentication; the first authentic
+MUST fail the request. If the server's current timestamp is past the `expiration-time` field, the server MUST fail the request. If the `expiration-time` field is
+greater than two weeks (1209600 seconds) ahead of the `issued-at` field, the server MUST fail the request. If all fields are acceptable, and the message
+is formatted correctly, the server MUST authenticate the user's session, and return an HTTP `200`. The server MUST attach the fields from the signed message to the session internally,
+in order to access data about the user and their session in other endpoints.
+
+Once a server authenticates a session using a particular nonce, the server MUST NOT allow any other sessions to be created using that nonce, until `expiration-time` of the session created
+using that nonce has been reached. The implementation of this is up to the server, but will require storing some global state about nonces in order to check if a client is attempting
+to sign-in using a nonce that is temporarily forbidden. Careful consideration should be taken to make sure that nonce management works correctly when scaling server resources.
+
+###### 3.2.1.3.3.2. Failure
+
+On failure, the server MUST return an HTTP `401` error, along with an error code that details exactly why the request failed.
+
+###### 3.2.1.3.3.2.1 `InvalidSignature`
+
+If the signature is invalid or doesn't match the included `address` field, the server MUST return an `InvalidSignature` error.
+
+###### 3.2.1.3.3.2.2 `InvalidParameters`
+
+If the request body is missing, malformed, or fails to pass any of the required checks, the server MUST respond with an `InvalidParameters` error.
+
+#### 3.2.1.4. Using Sessions
+
+Once a client has established a session with a server, they can use that session (by sending the session cookie to the server) to access priveleged endpoints
+throughout the API. Certain endpoints impose different requirements on the in-use session. There are *three levels* of endpoints, with respect to their authentication
+requirements. These are explained below.
+
+##### 3.2.1.4.1. Non-Privileged Endpoints
+
+The first group of endpoints are those that *do not* require the user to have a logged-in session. These are:
 
 * `GET /quote/in`
 * `GET /quote/out`
+
+Since it may be a painful user experience (or impossible) to have a user log in to a provider just to get a quote, authentication is not required for these endpoints. Of course,
+clients may still access these endpoitns when logged in, but a logged-in session (or any session at all) is not required to access these endpoints.
+
+##### 3.2.1.4.2. Privileged Endpoints
+
+The second, and largest set of endpoints, are the ones that require a user to be logged in. These endpoints are the ones that deal with reading and writing user-specific data, *except*
+those that initiate transfers:
+
+* `POST /kyc/:kycSchema`
```

Comment:
I have a question on how this auth fits into the recent Persona <> Fiatconnect planning I did. Specifically the situation where a user has not done persona yet. Here is the proposed flow:

> 1. User attempts to cash in/out and the Wallet sees that there is no KYC info for the FC parner
> 2. Wallet sends a POST to IHL saying that `[FcPartner]` needs `[KycSchema]` and IHL creates an entry in firebase for `[FcPartner]` and `[KycSchema]`.
> 3. User goes through Persona flow on their device
> 4. Webhook from Persona calls IHL, IHL updates the firebase entry for Persona with status updates, and once status === `'Approved'` calls POST /kyc/:kycSchema for all `[FcPartner]`'s in firebase that need KYC info from step 2.
> 5. Webhook from FC Patner calls IHL when the kyc status changes, IHL updates the firebase entry for `[FcPartner]`
> 6. Wallet calls IHL GET /kyc, gets the new status change, and the user can proceed with the transaction

How could IHL do step 4 with this auth?

Reply:
let's take this convo back to slack, getting kind of Valora-specific